af/analysis/rotation.py

```diff
@@ -0,0 +1,246 @@
+apt-key adv --verbose --keyserver hkp://keyserver.ubuntu.com --recv-keys 'B5A08F01796E7F521861B449372D1FF271F2DD50'
+apt-get update
+apt-get upgrade -qy
+echo > /etc/motd
+apt-get install -qy oohelperd
+--
+
+"""
+
+from configparser import ConfigParser
+from datetime import datetime
+import logging
+import random
+import sys
+import time
+
+import psycopg2  # debdeps: python3-psycopg2
+import statsd  # debdeps: python3-statsd
+import digitalocean  # debdeps: python3-digitalocean
+
+metrics = statsd.StatsClient("127.0.0.1", 8125, prefix="rotation")
+
+log = logging.getLogger("reprocessor")
+log.addHandler(logging.StreamHandler())  # Writes to console
+log.setLevel(logging.DEBUG)
+
+conffile_path = "/etc/ooni/rotation.conf"
+setup_script_path = "/etc/ooni/rotation_setup.sh"
+TAG = "roaming-th"
+
+
+def add_droplet_to_db_table(db_conn, dr):
+    vals = [dr.name, dr.region["slug"], dr.ip_address, dr.ip_v6_address]
+    q = """INSERT INTO test_helper_instances
+        (name, provider, region, ipaddr, ipv6addr)
+        VALUES (%s, 'Digital Ocean', %s, %s, %s)
+    """
+    with db_conn.cursor() as cur:
+        cur.execute(q, vals)
+        db_conn.commit()
+
+
+def drain_droplet_in_db_table(db_conn, dr):
+    q = """UPDATE test_helper_instances
+    SET draining_at = NOW()
+    WHERE name = %s AND provider = 'Digital Ocean' AND region = %s
+    """
+    vals = [dr.name, dr.region["slug"]]
+    with db_conn.cursor() as cur:
+        cur.execute(q, vals)
+        db_conn.commit()
+
+
+def delete_droplet_from_db_table(db_conn, dr):
+    q = """DELETE FROM test_helper_instances
+    WHERE name = %s AND provider = 'Digital Ocean' AND region = %s
+    """
+    vals = [dr.name, dr.region["slug"]]
+    with db_conn.cursor() as cur:
+        cur.execute(q, vals)
+        db_conn.commit()
+
+
+@metrics.timer("destroy_drained_droplets")
+def destroy_drained_droplets(db_conn, api, draining_time_minutes, live_droplets):
+    q = """SELECT name FROM test_helper_instances
+    WHERE provider = 'Digital Ocean'
+    AND draining_at IS NOT NULL
+    AND draining_at < NOW() - interval '%s minutes ago'
+    ORDER BY draining_at
+    """
+    v = [draining_time_minutes]
+    with db_conn.cursor() as cur:
+        cur.execute(q, v)
+        oldest = cur.fetchone()
+
+    if oldest:
+        oldest = oldest[0]
+    else:
+        log.info("No droplet to destroy")
+        return
+
+    to_delete = [d for d in live_droplets if d.name == oldest]
+    for droplet in to_delete:
+        log.info(f"Destroying {droplet.name} droplet")
+        delete_droplet_from_db_table(db_conn, droplet)
+        droplet.destroy()
+
+
+@metrics.timer("spawn_new_droplet")
+def spawn_new_droplet(api, dig_oc_token, live_regions, conf):
+    regions = set(r.slug for r in api.get_all_regions() if r.available is True)
+    preferred_regions = regions - live_regions
+    region = random.choice(list(preferred_regions or regions))
+
+    name = datetime.utcnow().strftime("roaming-th-%Y%m%d%H%M%S")
```

Comment:
Is there a chance for a race on the name of the droplet? Maybe for extra safety you can append a random hash to the end of it?

Reply:
No, starting a new droplet takes much longer than a second